crates/matrix-sdk/src/widget_api/handler/outgoing.rs

```diff
@@ -1,10 +1,63 @@
-use crate::widget_api::messages::{capabilities::Options, openid::State as OpenIDState};
+use crate::widget_api::messages::{
+    capabilities::Options,
+    openid::State as OpenIDState,
+    to_widget::{CapabilitiesUpdatedRequest, ToWidgetMessage},
+    Header, MessageBody,
+};
 
-use super::Request;
+use super::{Error, Result};
 
-#[allow(missing_debug_implementations)]
-pub enum Message {
-    SendMeCapabilities(Request<(), Options>),
-    CapabilitiesUpdated(Request<Options, ()>),
-    OpenIDUpdated(Request<OpenIDState, ()>),
+pub trait OutgoingMessage: Sized + Send + Sync + 'static {
+    type Response;
+
+    fn into_message(self, header: Header) -> ToWidgetMessage;
+    fn extract_response(message: ToWidgetMessage) -> Result<Self::Response>;
+}
+
+pub struct SendMeCapabilities;
+impl OutgoingMessage for SendMeCapabilities {
+    type Response = Options;
+
+    fn into_message(self, header: Header) -> ToWidgetMessage {
+        ToWidgetMessage::SendMeCapabilities(MessageBody::request(header, ()))
+    }
+
+    fn extract_response(msg: ToWidgetMessage) -> Result<Self::Response> {
+        match msg {
+            ToWidgetMessage::SendMeCapabilities(body) => Ok(body.response()?.capabilities),
+            _ => Err(Error::UnexpectedResponse),
+        }
+    }
+}
+
+pub struct CapabilitiesUpdated(pub CapabilitiesUpdatedRequest);
+impl OutgoingMessage for CapabilitiesUpdated {
+    type Response = ();
+
+    fn into_message(self, header: Header) -> ToWidgetMessage {
+        ToWidgetMessage::CapabilitiesUpdated(MessageBody::request(header, self.0))
+    }
+
+    fn extract_response(msg: ToWidgetMessage) -> Result<Self::Response> {
+        match msg {
+            ToWidgetMessage::CapabilitiesUpdated(body) => Ok(body.response()?),
+            _ => Err(Error::UnexpectedResponse),
+        }
+    }
+}
+
+pub struct OpenIDUpdated(pub OpenIDState);
+impl OutgoingMessage for OpenIDUpdated {
+    type Response = ();
+
+    fn into_message(self, header: Header) -> ToWidgetMessage {
+        ToWidgetMessage::OpenIdCredentials(MessageBody::request(header, self.0))
+    }
+
+    fn extract_response(msg: ToWidgetMessage) -> Result<Self::Response> {
+        match msg {
+            ToWidgetMessage::OpenIdCredentials(body) => Ok(body.response()?),
+            _ => Err(Error::UnexpectedResponse),
+        }
+    }
```

Comment:
This also screams for a macro. If I am not missing anything we basically only change the name and the `ToWidgetMessage::` enum and the way we cerate the `request`
```rust
outgoing_message!(
pub struct <MessageTypeName>(pub <BodyType>) -> response(<code to generate response body>), request(<code to generate request>: <ResposneType>)
)
```
^ probably you have a much better idea but this seems to be sufficient?

Reply:
Yes, indeed, those could also be simplified with a macro, I think I just prioritized improving it low compared to other parts so far.